### ISSUE TITLE
Code Coverage Reports in Deploy Commands Produce Unreliable `Covered` Lines

### ISSUE THREAD

**Issue by mcarvin8** [forcedotcom/cli#2771]
### Summary
This has already been reported as bugs such as https://github.com/forcedotcom/cli/issues/1568 but this has been closed due to "working as expected".

Please look into this soon for a new CLI Build.

When running `sf project deploy start`/`sf project deploy validate` commands with the code coverage flags (like `--coverage-formatters json`), the `uncovered` lines reported back are correct. The `covered` lines are incorrect for various reasons:
- `covered` lines will sometimes be out-of-range lines (if there is a 98 line Apex Class, the coverage reports will report line 100 as `covered`)
- `covered` lines are inconsistent and report back more lines than there are

Please update the Salesforce CLI to reliably report `covered` lines in the code coverage files so they can be exported to tools like SonarQube for code coverage tracking. In the current state, the `uncovered` lines can be exported to SonarQube correctly, but you need to have `covered` lines reported to SonarQube to calculate the coverage percentage. At this point, coverage in SonarQube will always be at 0% or the coverage % will be unreliable due to the `covered` lines being incorrectly reported.

I'm unclear if the CLI's `apex` commands produce the same outputs as the `deploy` commands in terms of code coverage. But we need reliable `deploy` coverage output so we can report back accurate code coverage during deployment and validation.

@xL3o - Please add anything I missed here based on our testing in my plugin.

See additional notes on my plugin here - https://github.com/mcarvin8/apex-code-coverage-transformer/issues/2

### Steps To Reproduce

1. Deploy a package with apex classes or triggers with the code coverage formatted as a JSON or JSON summary. `sf project deploy start -x manifest/package.xml --coverage-formatters json`
2. Confirm the covered and uncovered lines in the coverage JSON file created by the CLI

### Expected result
Uncovered and covered lines are reported back correctly. The covered lines should not have out-of-range line numbers.

### Actual result
Uncovered lines are correct. Covered Lines are incorrectly reported back.

### System Information
Tested on Windows 11/Powershell and Ubuntu/Bash (CI/CD).

```json
{
  "architecture": "win32-x64",
  "cliVersion": "@salesforce/cli/2.30.8",
  "nodeVersion": "node-v20.11.1",
  "osVersion": "Windows_NT 10.0.22621",
  "rootPath": "C:\\Users\\matthew.carvin\\AppData\\Local\\sf\\client\\2.30.8-d9ea14b",
  "shell": "cmd.exe",
  "pluginVersions": [
    "@oclif/plugin-autocomplete 3.0.9 (core)",
    "@oclif/plugin-commands 3.1.4 (core)",
    "@oclif/plugin-help 6.0.13 (core)",
    "@oclif/plugin-not-found 3.0.11 (core)",
    "@oclif/plugin-plugins 4.2.5 (core)",
    "@oclif/plugin-search 1.0.16 (core)",
    "@oclif/plugin-update 4.1.13 (core)",
    "@oclif/plugin-version 2.0.12 (core)",
    "@oclif/plugin-warn-if-update-available 3.0.12 (core)",
    "@oclif/plugin-which 3.1.0 (core)",
    "@salesforce/cli 2.30.8 (core)",
    "apex 3.0.25 (core)",
    "auth 3.3.13 (core)",
    "data 3.1.1 (core)",
    "deploy-retrieve 3.2.16 (core)",
    "dev 2.1.13 (user)",
    "info 3.0.27 (core)",
    "limits 3.1.10 (core)",
    "marketplace 1.0.25 (core)",
    "org 3.3.14 (core)",
    "packaging 2.1.10 (core)",
    "schema 3.1.3 (core)",
    "settings 2.0.27 (core)",
    "sobject 1.1.13 (core)",
    "source 3.1.15 (core)",
    "telemetry 3.1.13 (core)",
    "templates 56.0.16 (core)",
    "trust 3.3.10 (core)",
    "user 3.2.11 (core)",
    "sfdx-decomposer 3.1.0 (user)",
    "sfdx-git-delta 5.34.0 (user)"
  ]
}
```

**Comment by shetzel** [comment]
There is nothing new to report on the code coverage front from that last issue you mentioned.  The CLI team is dependent on some other internal teams and their priorities for any work to begin on this.  If the code coverage data in the API response is not accurate then there's nothing we can do except report what's given to us.  So from a CLI perspective, the code is making the correct calls and reporting the response as provided (i.e., working as designed), and to change that we need other teams to prioritize that work.  It's on their roadmaps but I don't know when it will be delivered.

**Comment by mcarvin8** [comment]
Should we keep this issue open @shetzel or just close it since this seems out of the CLI team's responsibility? Seems like once the API team updates the API response to provide accurate code coverage, it will work fine on any version of the CLI.

**Comment by mcarvin8** [comment]
@jfaderanga - Per @shetzel , this issue needs to be fixed by the team that manages the API first. I agree with you that this is a blocker for anyone who is trying to implement Quality Gates that enforce code coverage.

@shetzel - Is there a different way to contact the right team that owns updates to the API (outside of a general Salesforce support case)?

**Comment by shetzel** [comment]
I agree with both of you that it's an issue, just not an issue that anyone on the CLI team can do anything about.  This repo is about issues with the CLI.  If it's open, it implies we can take some action on it but in this case we can't.

>Is there a different way to contact the right team that owns updates to the API (outside of a general Salesforce support case)?

The best ways to influence the right team are opening a support case, and voting up the idea on the [IdeaExchange](https://ideas.salesforce.com/s/).

**Comment by mcarvin8** [comment]
> it is a blocker for pipeline with Quality Gates. I've been struggling with this issue for few months now. One thing I noticed is the code should be modularized and clean for the issue not to happened. There are times that the error still shows up even the code is modularized and well written which becomes a blocker on development as the code can't be enhanced anymore. 
> 
> 
> 
> Please re-open the issue. @mcarvin8 @shetzel 

https://github.com/mcarvin8/apex-code-coverage-transformer/pull/3

@jfaderanga - I was working on this with my plug-in that converts the coverage JSON to an XML SonarQube will accept. 

This will convert covered lines and work around this known issue by reassigning out of bound lines with unused lines. The coverage % will vary but it should still pass if you're hitting coverage. I could release this but I'm skeptical since the covered lines won't always report back the same way (comments and new lines can change it).

**Comment by mcarvin8** [comment]
@jfaderanga - My [plugin](https://github.com/mcarvin8/apex-code-coverage-transformer/blob/main/README.md) essentially just converts the JSON output from the deploy command into a XML file containing the uncovered lines/covered lines.
The latest build of my plugin just prints uncovered lines in the XML since that's accurate.
The beta build of my plugin prints covered and uncovered lines with some extra tinkering done on covered lines due to this bug.

All builds will take care of replacing `no-map` in the file name based on the `--dx-directory` flag (the default path being `force-app/main/default`.

**Comment by xL3o** [comment]
> The best ways to influence the right team are opening a support case, and voting up the idea on the [IdeaExchange](https://ideas.salesforce.com/s/).

@shetzel is IdeaExchange really the best way forward in our situation? I mean this code coverage functionality for validation deployments is already part of the end product, rather than being a new "Idea" that the POs would need to prioritize. Therefore, there should be a more streamlined approach how to highlight this issue/bug directly to the correct team for them to start bug fixing.

**Comment by shetzel** [comment]
@xL3o - I meant that more as a general statement of how to get issues in front of non-CLI teams.  For this issue, a support case is best.  However, I'll open this back up and transfer it to the apex-node library.  That team is a better choice to own this issue.
